### PR TITLE
(Develop Only) Fixes issue 5426: spurious warning for NVCC < 11.5 about missing return statement

### PR DESCRIPTION
**Pull request by crtrott**
See #5426

### CODE REVIEW

File: core/src/impl/Kokkos_ViewCtor.hpp [hunk 1]
```diff
@@ -267,6 +267,12 @@ auto with_properties_if_unset(const ViewCtorProp<P...> &view_ctor_prop) {
   return view_ctor_prop;
 }
 
+// Suppress faulty warning from NVCC see:
+// https://github.com/kokkos/kokkos/issues/5426
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1150)
+#pragma push
+#pragma diag_suppress = implicit_return_from_non_void_function
```
Comment (dalg24) [inline core/src/impl/Kokkos_ViewCtor.hpp:274]:
```suggestion
#pragma diag_suppress implicit_return_from_non_void_function
```

**Comment by crtrott** [comment]
So: the warning we are getting (and which is turned into an error) comes from GCC for an unknown pragma. I can go one step further and make this device code specific however ...

**Comment by dalg24** [comment]
> So: the warning we are getting (and which is turned into an error) comes from GCC for an unknown pragma. I can go one step further and make this device code specific however ...

Do you know why the warning was only emitted for one of the builds?

**Comment by crtrott** [comment]
Only one build which used NVCC older than 11.5 to build. The other ones are newer (CUDA 11.6/11.7) or use Clang as CUDA compiler.

File: core/src/Kokkos_Macros.hpp [hunk 2]
```diff
@@ -116,7 +116,11 @@
 // Code is parsed and separated into host and device code.
 // Host code is compiled again with another compiler.
 // Device code is compile to 'ptx'.
-#define KOKKOS_COMPILER_NVCC __NVCC__
+// NOTE: There is no __CUDACC_VER_PATCH__ officially, its __CUDACC_VER_BUILD__
+// which does have more than one digit (potentially undefined number of them).
+// This macro definition is in line with our other compiler defs
+#define KOKKOS_COMPILER_NVCC \
+  __CUDACC_VER_MAJOR__ * 100 + __CUDACC_VER_MINOR__ * 10
```
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp:123]: Is there a good reason to not define this as `CUDA_VERSION`?

File: core/src/impl/Kokkos_ViewCtor.hpp [hunk 3]
```diff
@@ -322,6 +332,13 @@ KOKKOS_FUNCTION const auto &get_property(
     return view_ctor_prop;
   }
 }
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1150)
+// pragma pop is getting a warning from the underlying GCC
+// for unknown pragma if -pedantic is used
+#ifdef __CUDAR_ARCH__
```
Comment (PhilMiller) [inline core/src/impl/Kokkos_ViewCtor.hpp:338]:
```suggestion
#ifdef __CUDA_ARCH__
```
Typo in the macro name

File: core/src/Kokkos_Macros.hpp (hunk 2 above)
Comment (crtrott) [reply core/src/Kokkos_Macros.hpp:123]: This is about which version of NVCC this is independent of whether we compile with clang or nvc++. Furthermore, I thought that KOKKOS_COMPILER_FOO is for most other compilers defined as FOO_VER_MAJOR*100+FOO_VER_MINOR*10+FOO_VER_PATCH.
Essentially there is a difference between checking for CUDA version and compiler version for the compiler you use to compile CUDA. Just for NVCC its kinda the same (except our fuckup with being inconsistent with KOKKOS_COMPILER_CUDA_VERSION)

File: core/src/impl/Kokkos_ViewCtor.hpp (hunk 3 above)
Comment (crtrott) [reply core/src/impl/Kokkos_ViewCtor.hpp]: gah ...

File: core/src/Kokkos_Macros.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:123]: Yeah, we should probably try to remove `KOKKOS_COMPILER_CUDA_VERSION`.

File: core/src/impl/Kokkos_ViewCtor.hpp [hunk 4]
```diff
@@ -267,6 +267,16 @@ auto with_properties_if_unset(const ViewCtorProp<P...> &view_ctor_prop) {
...
+// Suppress faulty warning from NVCC see:
+// https://github.com/kokkos/kokkos/issues/5426
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1150)
+// pragma pop is getting a warning from the underlying GCC
+// for unknown pragma if -pedantic is used
+#ifdef __CUDA_ARCH__
+#pragma push
+#pragma diag_suppress implicit_return_from_non_void_function
+#endif
+#endif
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewCtor.hpp:279]: I think this is better than trying to do any tag dispatch or SFINAE.

**Review by masterleinad — APPROVED** [review]
Thanks! Looks good to me.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**